synapse/rest/media/v1/media_storage.py

```diff
@@ -103,10 +102,9 @@ def store_into_file(self, file_info):
 
         finished_called = [False]
 
-        @defer.inlineCallbacks
-        def finish():
+        async def finish():
             for provider in self.storage_providers:
-                yield provider.store_file(path, file_info)
+                await provider.store_file(path, file_info)
```

Comment:
sadly I don't think we can rely on the storage providers honouring their interface. In particular, see https://github.com/matrix-org/synapse-s3-storage-provider/blob/master/s3_storage_provider.py#L96-L98: `reactor.callInThread` returns None, and `make_deferred_yieldable` passes it through.

in other words: better add some safety checks here.

Reply:
I added some checks in af4f079bf5ae5bec16868da61c4858b1c68bf672 for this, hopefully they make sense!